src/h2_stream_set.erl

```diff
@@ -0,0 +1,855 @@
+                {{#frame_header{
+                     stream_id=Stream#active_stream.id,
+                     type=?DATA,
+                     length=MaxToSend
+                    },
+                  http2_frame_data:new(BinToSend)},
+                 MaxToSend,
+                 Stream#active_stream{
+                   queued_data=Rest,
+                   send_window_size=SSWS-MaxToSend}}
+        end,
+
+    _Sent = http2_stream:send_frame(Stream#active_stream.pid, Frame),
+
+    case ExitStrategy of
+        max_frame_size ->
+            s_send_what_we_can(SWS - SentBytes, MFS, NewS);
+        stream ->
+            {SWS - SentBytes, NewS};
+        connection ->
+            {0, NewS}
+    end;
+s_send_what_we_can(SWS, _MFS, NonActiveStream) ->
+    {SWS, NonActiveStream}.
+
+
+%% Record Accessors
+-spec stream_id(
+        Stream :: stream()) ->
+                       stream_id().
+stream_id(#idle_stream{id=SID}) ->
+    SID;
+stream_id(#active_stream{id=SID}) ->
+    SID;
+stream_id(#closed_stream{id=SID}) ->
+    SID.
+
+-spec pid(stream()) -> pid() | undefined.
+pid(#active_stream{pid=Pid}) ->
+    Pid;
+pid(_) ->
+    undefined.
+
+-spec type(stream()) -> idle | active | closed.
+type(#idle_stream{}) ->
+    idle;
+type(#active_stream{}) ->
+    active;
+type(#closed_stream{}) ->
+    closed.
+
+queued_data(#active_stream{queued_data=QD}) ->
+    QD;
+queued_data(_) ->
+    undefined.
+
+update_data_queue(
+  NewBody,
+  BodyComplete,
+  #active_stream{} = Stream) ->
+    Stream#active_stream{
+      queued_data=NewBody,
+      body_complete=BodyComplete
+     };
+update_data_queue(_, _, S) ->
+    S.
+
+response(#closed_stream{
+            response_headers=Headers,
+            response_body=Body}) ->
+    {Headers, Body};
+response(_) ->
+    no_response.
+
+recv_window_size(#active_stream{recv_window_size=RWS}) ->
+    RWS;
+recv_window_size(_) ->
+    undefined.
+
+decrement_recv_window(
+  L,
+  #active_stream{recv_window_size=RWS}=Stream
+ ) ->
+    Stream#active_stream{
+      recv_window_size=RWS-L
+     };
+decrement_recv_window(_, S) ->
+    S.
+
+send_window_size(#active_stream{send_window_size=SWS}) ->
+    SWS;
+send_window_size(_) ->
+    undefined.
+
+increment_send_window_size(
+  WSI,
+  #active_stream{send_window_size=SWS}=Stream) ->
+    Stream#active_stream{
+      send_window_size=SWS+WSI
+     };
+increment_send_window_size(_WSI, Stream) ->
+    Stream.
+
+stream_pid(#active_stream{pid=Pid}) ->
+    Pid;
+stream_pid(_) ->
+    undefined.
+
+%% the false clause is here as an artifact of us using a simple
+%% lists:keyfind
+notify_pid(#idle_stream{}) ->
+    undefined;
+notify_pid(#active_stream{notify_pid=Pid}) ->
+    Pid;
+notify_pid(#closed_stream{notify_pid=Pid}) ->
+    Pid.
+
+my_active_count(SS) ->
+    SS#stream_set.mine#peer_subset.active_count.
+
+their_active_count(SS) ->
+    SS#stream_set.theirs#peer_subset.active_count.
+
+my_active_streams(SS) ->
```

Comment:
could you add a comment / docstring saying this?

On Thu, Apr 28, 2016 at 2:31 PM Joe DeVivo <notifications@github.com> wrote:

> In src/h2_stream_set.erl
> <https://github.com/joedevivo/chatterbox/pull/48#discussion_r61478820>:
>
> > +%% the false clause is here as an artifact of us using a simple
> > +%% lists:keyfind
> > +notify_pid(#idle_stream{}) ->
> > +    undefined;
> > +notify_pid(#active_stream{notify_pid=Pid}) ->
> > +    Pid;
> > +notify_pid(#closed_stream{notify_pid=Pid}) ->
> > +    Pid.
> > +
> > +my_active_count(SS) ->
> > +    SS#stream_set.mine#peer_subset.active_count.
> > +
> > +their_active_count(SS) ->
> > +    SS#stream_set.theirs#peer_subset.active_count.
> > +
> > +my_active_streams(SS) ->
>
> _count is the number of #active_stream{}s, _active is the actual working
> set of #active_stream{}s and #closed_streams{}s that have responses that
> haven't been gc'd
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joedevivo/chatterbox/pull/48/files/6dee19685846c614057cc5068cf3796ba9bbf1e4#r61478820>
>